c/parallel/src/unique_by_key.cu

```diff
@@ -0,0 +1,514 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#include <cub/detail/choose_offset.cuh>
+#include <cub/detail/launcher/cuda_driver.cuh>
+#include <cub/device/device_select.cuh>
+
+#include <format>
+
+#include "kernels/iterators.h"
+#include "kernels/operators.h"
+#include "util/context.h"
+#include "util/indirect_arg.h"
+#include "util/scan_tile_state.h"
+#include "util/types.h"
+#include <cccl/c/unique_by_key.h>
+#include <nvrtc/command_list.h>
+#include <nvrtc/ltoir_list_appender.h>
+
+struct op_wrapper;
+struct device_unique_by_key_policy;
+using OffsetT = int64_t;
+static_assert(std::is_same_v<cub::detail::choose_signed_offset_t<OffsetT>, OffsetT>, "OffsetT must be int64");
+
+struct input_keys_iterator_state_t;
+struct input_values_iterator_state_t;
+struct output_keys_iterator_t;
+struct output_values_iterator_t;
+struct output_num_selected_iterator_t;
+
+struct num_selected_storage_t;
+
+namespace unique_by_key
+{
+struct unique_by_key_runtime_tuning_policy
+{
+  int block_size;
+  int items_per_thread;
+  cub::BlockLoadAlgorithm load_algorithm;
+  cub::CacheLoadModifier load_modifier;
+
+  unique_by_key_runtime_tuning_policy UniqueByKey() const
+  {
+    return *this;
+  }
+
+  using UniqueByKeyPolicyT = unique_by_key_runtime_tuning_policy;
+};
+
+struct unique_by_key_tuning_t
+{
+  int cc;
+  int block_size;
+  int items_per_thread;
+};
+
+template <typename Tuning, int N>
+Tuning find_tuning(int cc, const Tuning (&tunings)[N])
+{
+  for (const Tuning& tuning : tunings)
+  {
+    if (cc >= tuning.cc)
+    {
+      return tuning;
+    }
+  }
+
+  return tunings[N - 1];
+}
+
+unique_by_key_runtime_tuning_policy get_policy(int /*cc*/)
+{
+  // TODO: we should update this once we figure out a way to reuse
+  // tuning logic from C++. Alternately, we should implement
+  // something better than a hardcoded default:
+  return {128, 10, cub::BLOCK_LOAD_DIRECT, cub::LOAD_DEFAULT};
```

Comment:
important: we should wait with copy-pasting all the tuning policies we have on the C++ end until we have a machinery in place. That said, unique by key tuning uses items per thread scaling to better handle large data types. It shouldn't be a lot of effort to make the tuning policy depend on the key (similar to `Nominal4BItemsToItems<KeyT>(10)` logic) type along the lines of what we do in the reduce.cu. But if it's not the case, feel free to address this as a follow-up PR.

Reply:
I had original implemented `Nominal4BItemsToItems` but that triggered the memory op assertion in the tests, so I overrode that value it produced with a smaller one. I moved `Nominal4BItemsToItems` to a common file to use it here, but I ran into the same issue, so I fixed it the same way. I think I'm just going to fix both in a later PR

Reply:
Created https://github.com/NVIDIA/cccl/issues/4013